### PR TITLE
Added some more tests

### PR DESCRIPTION
**Pull request by codeinred**
I added some more tests to the tests/ directory! There are two new tests: 

- A read-write test, where data is read, and then written - this should fail if reading blocks
- A ping-pong test. Two coroutines send messages back and forth, each time reading the response of the other. I think it has some fun output!

Both tests pass.
image

image

### CODE REVIEW

File: tests/ping_pong.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+}
+auto pong(uio::io_service& service, int read_fd, int write_fd) -> uio::task<> {
+    std::string_view ping = "pong!";
+    std::array<char, 64> buffer;
+    for (int i = 0; i < 20; i++) {
+        co_await service.write(write_fd, ping.data(), ping.size(), 0);
+        co_await service.fsync(write_fd, 0);
+
+        int count =
+            co_await service.read(read_fd, buffer.data(), buffer.size(), 0);
```
Comment (CarterLi) [inline tests/ping_pong.cpp:38]: io_uring functions don't set the global errno, they return -errno on error. See panic_on_error

File: tests/read_write_pipe.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#include <liburing/io_service.hpp>
+#include <string_view>
+
+#include <fmt/core.h>
+
+auto write_to_fd(uio::io_service& service, int fd) -> uio::task<> {
+    std::string_view message = "Hello, world!";
+    fmt::print("Sending data...\n");
+    co_await service.write(fd, message.data(), message.size(), 0)
```
Comment (CarterLi) [inline tests/read_write_pipe.cpp:9]: These two tests seems to be similar. You may try something different

File: tests/ping_pong.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        if (count < 0)
+            uio::panic("Bad read", errno);
```
Comment (CarterLi) [inline tests/ping_pong.cpp:40]: io_uring IO functions don't set the global errno, they return -errno on error. See panic_on_error

File: tests/ping_pong.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
+#include <fmt/color.h>
+#include <fmt/core.h>
+
+#include <liburing/io_service.hpp>
+#include <string_view>
+auto ping(uio::io_service& service, int read_fd, int write_fd) -> uio::task<> {
```
Comment (CarterLi) [inline tests/ping_pong.cpp:6]: A newline is required after `#include`s

File: tests/ping_pong.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            fg(fmt::color::magenta) | fmt::emphasis::bold,
+            "pong: Recieved {}\n",
+            recieved);
+        if (recieved != "ping!")
+            uio::panic("Unexpected message", 0);
+    }
+    close(read_fd);
+    close(write_fd);
+}
+int main() {
```
Comment (CarterLi) [inline tests/ping_pong.cpp:54]: Newline between functions please

File: tests/read_write_pipe.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    std::array<int, 2> fds;
+    pipe(fds.data()) | uio::panic_on_err("Unable to open pipe", true);
+
+    auto t1 = read_from_fd(io, fds[0]);
+    auto t2 = write_to_fd(io, fds[1]);
+
+    io.run(t1);
+    io.run(t2);
+}
```
Comment (CarterLi) [inline tests/read_write_pipe.cpp:50]: ping_pong.cpp and read_write_pipe.cpp test the same functions ( read / write / pipe ), please try something different.

File: tests/ping_pong.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        co_await service.write(write_fd, ping.data(), ping.size(), 0);
+        co_await service.fsync(write_fd, 0);
+    }
+    close(read_fd);
+    close(write_fd);
+}
+auto pong(uio::io_service& service, int read_fd, int write_fd) -> uio::task<> {
+    std::string_view ping = "pong!";
+    std::array<char, 64> buffer;
+    for (int i = 0; i < 20; i++) {
```
Comment (CarterLi) [inline tests/ping_pong.cpp:33]: As pong is a receiver, you may test for EOF here

Comment (codeinred) [reply tests/ping_pong.cpp:33]: Just to confirm: the value returned from `service.read()` 0 on EOF?

Comment (CarterLi) [reply tests/ping_pong.cpp:33]: `<= 0` should be ok

**Comment by codeinred** [comment]
Hi Carter,

I updated tests/ping_pong.cpp in accordance with the formatting guidelines, and I also fixed the error handling code.

> ping_pong.cpp and read_write_pipe.cpp test the same functions ( read / write / pipe ), please try something different.

Regarding this statement, I decided to remove *read_write_pipe* since ping_pong handled it more thoroughly and is a more interesting test. Please let me know if there's anything else you need!

**Comment by CarterLi** [comment]
Merged, thanks.

**Pull request merged**